File: objective-c/oauthdemoapp/ViewController.m
//
//  ViewController.m
//  oauthdemoapp
//

#import "ViewController.h"
#import "ServerManager.h"
#import "Token.h"

#define kMainQueue dispatch_get_main_queue()
#define MY_CUSTOM_TAG 1234

@interface ViewController ()

@property (strong, nonatomic) NSArray *demoDataArray;
@property (strong, nonatomic) UIBarButtonItem *loginButton;
@property (strong, nonatomic) Token *token;

@end


@implementation ViewController

- (id)init {
    self = [super init];
    if (self) {
        self.token = [Token new];
        self.demoDataArray = [[NSArray alloc] init];
        self.tableView.tag = MY_CUSTOM_TAG;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // add login button
    self.loginButton = [[UIBarButtonItem alloc] initWithTitle:@"login"
                                                        style:UIBarButtonItemStyleDone
                                                       target:self
                                                       action:@selector(loginButtonPressed:)];
    [self.navigationItem setLeftBarButtonItem:self.loginButton animated:YES];
    
    // set title
    self.title = @"OAuthApp demo";
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


#pragma mark - Actions

- (void)loginButtonPressed:(UIButton *)button {
    [[ServerManager sharedInstance] authorizeUser:^(NSArray *data) {
        self.demoDataArray = data;
        NSLog(@"%@", self.demoDataArray);
        
        dispatch_async(kMainQueue, ^{
            [self.tableView reloadData];
        });
    }];
}


#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.demoDataArray count];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *identifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
    }

    cell.textLabel.text = [[self.demoDataArray objectAtIndex:indexPath.row] objectForKey:@"FriendlyDesc2"];
    
    return cell;
}




@end

File: swift/oauthdemoapp-swift/ViewController.swift
//
//  ViewController.swift
//

import UIKit

class ViewController: UITableViewController {
    
    var demoDataArray: NSArray = NSArray()
    var loginButton: UIBarButtonItem?
    var token: Token = Token()

    
    override func viewDidLoad() {
        super.viewDidLoad()
        let homeButton: UIBarButtonItem = UIBarButtonItem(title: "login", style: UIBarButtonItemStyle.Done, target: self, action: "loginButtonPressed:")
        self.navigationItem.setLeftBarButtonItem(homeButton, animated: true)
        self.title = "OAuth demo application"
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: Actions
    func loginButtonPressed(sender: UIButton) -> Void {
        ServerManager.sharedInstance.authorizeUser { (demoData) -> Void in
            if demoData != nil {
                self.demoDataArray = demoData!
                // print(self.demoDataArray)
                dispatch_async(dispatch_get_main_queue(), { () -> Void in
                    var indexPath:[NSIndexPath] = [NSIndexPath]()
                    for var i = 0; i < self.demoDataArray.count; i++ {
                        indexPath.append(NSIndexPath(forRow: i, inSection: 0))
                    }
                    self.tableView.beginUpdates()
                    self.tableView.insertRowsAtIndexPaths(indexPath, withRowAnimation: UITableViewRowAnimation.Top)
                    self.tableView.endUpdates()
                })
            }
        }
    }
    
    // MARK: UITableViewDataSource
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        if let demoText = self.demoDataArray.objectAtIndex(indexPath.row) as? NSDictionary {
            let text = DemoDataCell.prepareText(demoText)
            let height = DemoDataCell.heightForRow(text)
            return height
        } else {
            return CGFloat()
        }
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.demoDataArray.count
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath) as! DemoDataCell
        if let demoText = self.demoDataArray.objectAtIndex(indexPath.row) as? NSDictionary {
            let text = DemoDataCell.prepareText(demoText)
            cell.demoTextLable.text = text
            cell.demoTextLable.lineBreakMode = NSLineBreakMode.ByWordWrapping
        }
        return cell
    }

}

